lametro/views.py

```diff
@@ -379,54 +378,52 @@ def day_grouper(x):
-                org_select_events.append([date(*event_date), events])
-
-            context["select_events"] = org_select_events
+            org_select_events = itertools.groupby(select_events, key=day_grouper)
+            context["select_events"] = [
+                (d, list(events)) for d, events in org_select_events
+            ]
 
-        # If all meetings
-        elif self.request.GET.get("show"):
+        # Did user request all events?
+        elif self.request.GET.get("show") == "all":
             all_events = media_events.order_by("-start_time")
-            org_all_events = []
+            org_all_events = itertools.groupby(all_events, key=day_grouper)
 
-            for event_date, events in itertools.groupby(all_events, key=day_grouper):
-                events = sorted(events, key=attrgetter("start_time"))
-                org_all_events.append([date(*event_date), events])
+            context["all_events"] = [(d, list(events)) for d, events in org_all_events]
 
-            context["all_events"] = org_all_events
-        # If no...
         else:
             # Upcoming events
             future_events = media_events.filter(start_time__gt=timezone.now()).order_by(
                 "start_time"
             )
-            org_future_events = []
-
-            for event_date, events in itertools.groupby(future_events, key=day_grouper):
-                events = sorted(events, key=attrgetter("start_time"))
-                org_future_events.append([date(*event_date), events])
-
-            context["future_events"] = org_future_events
+            org_future_events = itertools.groupby(future_events, key=day_grouper)
 
             # Past events
-            past_events = media_events.filter(start_time__lt=timezone.now()).order_by(
-                "-start_time"
+            past_events = (
+                media_events.filter(start_time__lt=timezone.now())
+                .order_by("-start_time")
+                .prefetch_related(
+                    Prefetch("documents", minutes_queryset, to_attr="minutes")
+                )
+                .prefetch_related("minutes__links")
             )
-
-            past_events = past_events.prefetch_related(
-                Prefetch("documents", minutes_queryset, to_attr="minutes")
-            ).prefetch_related("minutes__links")
-
-            org_past_events = []
-
-            for event_date, events in itertools.groupby(past_events, key=day_grouper):
-                events = sorted(events, key=attrgetter("start_time"))
-                org_past_events.append([date(*event_date), events])
-
-            context["past_events"] = org_past_events
+            org_past_events = itertools.groupby(past_events, key=day_grouper)
+
+            # Only show three days of events by default
+            past_events = itertools.islice(org_past_events, 3)
+            future_events = itertools.islice(org_future_events, 3)
```

Comment:
Hmm, rather than retrieving all events, grouping, and the limiting to three days, what if we filtered the queryset for events on the most recent past three days? Something like:

```python
media_events.filter(start_date__in=LAMetroEvent.objects.filter(
    start_time__lt=timezone.now()
).order_by("-start_time")[:3])
```

And then applying grouping?